Reload web browser page when firmware version change detected

chriadam opened the pull request:
If the device has its firmware updated, it may have a new build of the UI included.  To ensure that the version of the UI which is being used via the browser matches the version of the UI which is being used on device, we need to reload the page so that the up-to-date wasm binary is reloaded.

jpetrell commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,74 @@ QtObject {
 	}
 
+	property VeQuickItem _firmwareInstalledBuild: VeQuickItem {
+		property string newValue
```
newValue seems unnecessary? it is only used within one signal handler?

chriadam replied:
fixed

jpetrell commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,74 @@ QtObject {
+		onValueChanged: {
+			if (value != undefined) {
+				newValue = value
+				if (newValue.length > 0) {
```
If newValue gets dropped could combine the conditionals `if (value != undefined && value.length > 0) { ..`

jpetrell commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,74 @@ QtObject {
+						} else {
+							timer.needRestart = true
+							timer.interval = 2000
+							timer.start()
```
Would consider creating another Timer instead to improve code readability.

chriadam replied:
done

jpetrell commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,74 @@ QtObject {
+			property bool needRestart
+			interval: 10 * 60 * 1000
+			repeat: true
+			running: true
```
running: Qt.platform.os == "wasm"?

chriadam replied:
yes, done

jpetrell commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,74 @@ QtObject {
+
+		property Component firmwareVersionRestartDialog: Component {
+			ModalWarningDialog {
+				id: dlg
```
Unused id. I generally take visible id declaration as an indication that the element is accessed somewhere.

chriadam replied:
indeed, removed

jpetrell commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,74 @@ QtObject {
+				//% "Firmware version update detected"
+				title: qsTrId("firmware_installed_build_version_update_detected")
+				//% "Page will automatically reload in ten seconds to load the latest version"
+				description: qsTrId("firmware_installed_build_page_will_reload")
```
Is there a rule when to end the sentences in period? Anyhow most ModalWarningDialog I could find have titles that don't use periods, but descriptions that do.

ReinvdZee reviewed (approved):
LGTM

blammit reviewed (approved):
Can you add the task number to the commit?

blammit commented:
@chriadam can you remind me why the  `/Firmware/Installed/Build` value check is needed? If I update the firmware on the cerbo, the UI in wasm will trigger the "Firmware installed, rebooting" toast notification from https://github.com/victronenergy/gui-v2/blob/2a28d7b0ebc2b3893bf8a5f631ff9e4f7c3fbf62/components/FirmwareUpdate.qml#L57. Why can't we just trigger the reboot dialog from there on wasm, instead of having this separate `/Firmware/Installed/Build` check? Are there cases where that `FirmwareUpdate` state change is not triggered when the firmware is updated?

chriadam commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,70 @@ QtObject {
+	property VeQuickItem _firmwareInstalledBuild: VeQuickItem {
+		uid: Global.venusPlatform.serviceUid + "/Firmware/Installed/Build"
+		onValueChanged: {
+			if (value != undefined && value != null && value.length > 0) {
```
indeed, did null and length>0

chriadam commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,70 @@ QtObject {
+		uid: Global.venusPlatform.serviceUid + "/Firmware/Installed/Build"
+		onValueChanged: {
+			if (value != undefined && value != null && value.length > 0) {
+				console.log("Firmware version received: " + value + ", previously seen value: " + Global.firmwareInstalledBuild)
```
true, removed

chriadam commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,70 @@ QtObject {
+		}
+
+		property Timer forceReadTimer: Timer {
+			property bool needRestart
```
removed

chriadam commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,70 @@ QtObject {
+			if (value != undefined && value != null && value.length > 0) {
+				console.log("Firmware version received: " + value + ", previously seen value: " + Global.firmwareInstalledBuild)
+				if (Qt.platform.os == "wasm"
+						&& Global.firmwareInstalledBuild.length > 0
```
It is needed, because on first start it will initially be empty.  This is fine, and doesn't warrant a reboot, thus this forms part of the condition.

chriadam commented on data/SystemSettings.qml:
```diff
@@ -278,6 +278,70 @@ QtObject {
+						+ " (" + Global.firmwareInstalledBuild + " != " + value + ")")
+					// If the value change was only detected after a reconnection (and rebuildUi)
+					// then the dialogLayer may not have been constructed yet.
+					// ApplicationContent is constructed asynchronously, so might require retries.
```
Good points, done.

chriadam commented:
> Can you add the task number to the commit?

Done

chriadam commented:
> @chriadam can you remind me why the `/Firmware/Installed/Build` value check is needed? If I update the firmware on the cerbo, the UI in wasm will show the "Firmware installed, rebooting" toast notification due to this line in `FirmwareUpdate`:
> 
> https://github.com/victronenergy/gui-v2/blob/2a28d7b0ebc2b3893bf8a5f631ff9e4f7c3fbf62/components/FirmwareUpdate.qml#L57
> 
> Why can't we just trigger the reboot dialog from there on wasm, instead of having this separate `/Firmware/Installed/Build` check? Are there cases where that `FirmwareUpdate` state change is not triggered when the firmware is updated?

I've moved the impl to the FirmwareUpdate - that's definitely where it belongs.
We do still need the Firmware/Installed/Build check, though, for the case where we do a downgrade / reboot previously installed firmware version.
I've added a comment about why we don't want to trigger page reload immediately upon detecting that the device is rebooting.

chriadam commented on pages/DialogLayer.qml:
```diff
@@ -9,13 +9,34 @@ import Victron.VenusOS
+
 	function open(dialogComponent, properties) {
-		const dialog = dialogComponent.createObject(Global.dialogLayer, properties)
+		const dialog = dialogComponent.createObject(root, properties)
```
this change is because Global.dialogLayer won't be set until the Component.onCompleted handler is run, which might not happen until after the property change is noticed and reacted to.  I could have called Qt.callLater() in on_NeedPageReloadChanged, but I think it's more direct to simply specify the parent properly/directly here in this function (since Global.dialogLayer should always equal root anyway).
